pulley/src/interp.rs

```diff
@@ -597,9 +596,14 @@ impl MachineState {
     }
 }
 
-enum Continuation {
-    Continue,
+/// The reason the interpreter loop terminated.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+enum Done {
+    /// A `ret` instruction was executed and the call stack was empty. This is
+    /// how the loop normally ends.
     ReturnToHost,
+
+    /// a `trap` instruction was executed.
     Trap,
```

Comment:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4536593481e0f790a4f2d31c969b7e19

`ControlFlow<Done, ()>` and `Continuation` are both 1 byte (it seems rustc is smart enough to use the "unused discriminants" from `Done` as niches for `ControlFlow`).

`ControlFlow<Done, OpcodeHandler>` is 2 words. The alternative would be to return `ControlFlow<Done, ()>` and set the next handler via an `&mut OpcodeHandler` parameter, but I expect that might be harder for LLVM to optimize. In the worst case, returning a large aggregate will be lowered to returning via an out-parameter anyway, so it shouldn't be any worse performance wise

Reply:
We could also squeeze `ControlFlow<Done, OpcodeHandler>` into a single word by using the lower 2 bits as the discriminant tag, since function pointers must be word aligned. I think there was a proposal for rustc to do this automatically but it has not yet been implemented.